File: arena-xunit/tests/Unit/LocalstackDependencyBuilderSerializationTest.cs
using System;
using System.Linq;
using ArenaDotnet.Xunit.Dep;
using ArenaDotnet.Xunit.Support;
using Newtonsoft.Json.Linq;
using Xunit;

namespace ArenaDotnet.Xunit.UnitTest;

public class LocalstackDependencyBuilderSerializationTest
{
    [Fact]
    public void Build_DefaultPort_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("localstack", obj["type"]);
        Assert.Equal(4566, obj["port"]);
        Assert.NotNull(obj["identifier"]);
    }

    [Fact]
    public void Build_CustomPort_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithPort(4577).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal(4577, obj["port"]);
    }

    [Fact]
    public void Build_WithService_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithService("sqs").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var services = obj["services"];
        Assert.NotNull(services);
        Assert.Equal("sqs", services[0]);
    }

    [Fact]
    public void Build_WithServices_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithServices("sqs", "s3").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var services = obj["services"];
        Assert.NotNull(services);
        Assert.Equal(2, services.Count());
    }

    [Fact]
    public void Build_WithQueue_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithQueue("myqueue").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var queues = obj["queues"];
        Assert.NotNull(queues);
        var queue = Assert.Single(queues);
        Assert.Equal("myqueue", queue["name"]);
        Assert.Equal(false, queue["fifo"]);
    }

    [Fact]
    public void Build_WithFifoQueue_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithFifoQueue("myqueue.fifo").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var queues = obj["queues"];
        Assert.NotNull(queues);
        var queue = Assert.Single(queues);
        Assert.Equal("myqueue.fifo", queue["name"]);
        Assert.Equal(true, queue["fifo"]);
    }

    [Fact]
    public void Build_WithEventBus_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithEventBus("my-bus").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var eventBuses = obj["event_buses"];
        Assert.NotNull(eventBuses);
        var eventBus = Assert.Single(eventBuses);
        Assert.Equal("my-bus", eventBus["name"]);
    }

    [Fact]
    public void Build_WithImageName_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithImageName("localstack/localstack-pro").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("localstack/localstack-pro", obj["image_name"]);
    }

    [Fact]
    public void Build_WithImageTag_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithImageTag("3.8").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("3.8", obj["image_tag"]);
    }

    [Fact]
    public void Build_WithContainerName_SerializesCorrectJson()
    {
        var dep = new LocalstackDependencyBuilder("test").WithContainerName("my-container").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("my-container", obj["container_name"]);
    }

    [Fact]
    public void Build_EndpointUrl_Correct()
    {
        var dep = new LocalstackDependencyBuilder("test").WithPort(4577).Build();
        Assert.Equal("http://localhost:4577", dep.EndpointUrl);
    }

    [Fact]
    public void Build_Identifier_MatchesPattern()
    {
        var dep = new LocalstackDependencyBuilder("test").Build();
        Assert.StartsWith("arena-localstack-", dep.Identifier);
    }

    [Fact]
    public void Build_WithEventRuleSqsTarget_SerializesCorrectJson()
    {
        var target = EventRuleTargetBuilder.SqsQueue("t1", "queue1");
        var rule = new EventRuleSpec
        {
            Name = "rule1",
            EventPattern = "{}",
            Targets = new System.Collections.Generic.List<EventRuleTarget> { target }
        };
        var dep = new LocalstackDependencyBuilder("test").WithEventRule(rule).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var eventRules = obj["event_rules"];
        Assert.NotNull(eventRules);
        var eventRule = Assert.Single(eventRules);
        Assert.Equal("rule1", eventRule["name"]);
        var targets = eventRule["targets"];
        Assert.NotNull(targets);
        var eventTarget = Assert.Single(targets);
        Assert.Equal("sqs_queue", eventTarget["kind"]);
    }

    [Fact]
    public void Build_WithLambda_SerializesCorrectJson()
    {
        var spec = new LambdaSpec(
            "my-fn",
            "python3.12",
            "handler.main",
            ".",
            new[] { new System.Collections.Generic.KeyValuePair<string, string>("KEY", "VALUE") });
        var dep = new LocalstackDependencyBuilder("test").WithLambda(spec).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var lambdas = obj["lambdas"];
        Assert.NotNull(lambdas);
        var lambda = Assert.Single(lambdas);
        Assert.Equal("my-fn", lambda["name"]);
        Assert.Equal("python3.12", lambda["runtime"]);
        Assert.Equal("handler.main", lambda["handler"]);
        Assert.True(System.IO.Path.IsPathRooted((string)lambda["source_dir"]!));
        var environment = lambda["environment"];
        Assert.NotNull(environment);
        var pair = Assert.Single(environment);
        Assert.Equal("KEY", pair[0]);
        Assert.Equal("VALUE", pair[1]);
    }

    [Fact]
    public void Build_WithLambdaHomeRelativeSourceDir_ExpandsToUserProfile()
    {
        var spec = new LambdaSpec("my-fn", "python3.12", "handler.main", "~/my-lambda-src");
        var dep = new LocalstackDependencyBuilder("test").WithLambda(spec).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var lambda = Assert.Single(obj["lambdas"]!);
        var sourceDir = (string)lambda["source_dir"]!;
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        Assert.StartsWith(home, sourceDir);
        Assert.DoesNotContain("~", sourceDir);
    }

    [Fact]
    public void Build_WithEventRuleLambdaTarget_SerializesCorrectJson()
    {
        var target = EventRuleTargetBuilder.Lambda("t2", "func1");
        var rule = new EventRuleSpec
        {
            Name = "rule2",
            EventBus = "my-bus",
            EventPattern = "{}",
            Targets = new System.Collections.Generic.List<EventRuleTarget> { target }
        };
        var dep = new LocalstackDependencyBuilder("test").WithEventRule(rule).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        var eventRules = obj["event_rules"];
        Assert.NotNull(eventRules);
        var eventRule = Assert.Single(eventRules);
        var targets = eventRule["targets"];
        Assert.NotNull(targets);
        var eventTarget = Assert.Single(targets);
        Assert.Equal("lambda", eventTarget["kind"]);
        Assert.Equal("func1", eventTarget["function_name"]);
    }
}

File: arena-junit/src/test/java/arena/junit/dep/LocalstackDependencyBuilderSerializationTest.java
package arena.junit.dep;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import arena.junit.match.ArenaRunnableDependency;
import arena.junit.playbook.LocalstackModels;
import arena.junit.support.ArenaJson;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;
import org.junit.jupiter.api.Test;

final class LocalstackDependencyBuilderSerializationTest {

  static final class StubDependency implements ArenaRunnableDependency {
    private final String identifier;

    StubDependency(String identifier) {
      this.identifier = identifier;
    }

    @Override
    public ObjectNode forFfi() {
      ObjectNode n = ArenaJson.object();
      n.put("identifier", identifier);
      return n;
    }
  }

  @Test
  void build_minimalName_serializesTypeAndEmptyCollections() {
    ObjectNode config = new LocalstackDependencyBuilder("ls").build().forFfi();
    assertEquals("localstack", config.path("type").asText());
    assertTrue(config.path("identifier").asText().startsWith("arena-localstack-ls-"));
    assertTrue(config.path("services").isEmpty());
    assertTrue(config.path("queues").isEmpty());
    assertTrue(config.path("lambdas").isEmpty());
    assertTrue(config.path("event_buses").isEmpty());
    assertTrue(config.path("event_rules").isEmpty());
    assertFalse(config.has("children"));
  }

  @Test
  void withPortImageAndContainerName_setsScalarFields() {
    ObjectNode config =
        new LocalstackDependencyBuilder("ls")
            .withPort(4567)
            .withImageName("localstack/localstack")
            .withImageTag("3.0")
            .withContainerName("ls-box")
            .build()
            .forFfi();
    assertEquals(4567, config.path("port").asInt());
    assertEquals("localstack/localstack", config.path("image_name").asText());
    assertEquals("3.0", config.path("image_tag").asText());
    assertEquals("ls-box", config.path("container_name").asText());
  }

  @Test
  void withServices_singleAndBulk_appendsToServicesArray() {
    ObjectNode config =
        new LocalstackDependencyBuilder("ls")
            .withService("sqs")
            .withServices(List.of("sns", "lambda"))
            .build()
            .forFfi();
    ArrayNode services = (ArrayNode) config.path("services");
    assertEquals(3, services.size());
    assertEquals("sqs", services.get(0).asText());
    assertEquals("sns", services.get(1).asText());
    assertEquals("lambda", services.get(2).asText());
  }

  @Test
  void withQueueAndFifoQueue_appendsQueueSpecs() {
    ObjectNode config =
        new LocalstackDependencyBuilder("ls")
            .withQueue("plain-queue")
            .withFifoQueue("ordered-queue")
            .build()
            .forFfi();
    ArrayNode queues = (ArrayNode) config.path("queues");
    assertEquals(2, queues.size());
    assertEquals("plain-queue", queues.get(0).path("name").asText());
    assertFalse(queues.get(0).path("fifo").asBoolean());
    assertEquals("ordered-queue", queues.get(1).path("name").asText());
    assertTrue(queues.get(1).path("fifo").asBoolean());
  }

  @Test
  void withLambda_absoluteSourceDirAndEnvironment_serializesLambdaSpec() {
    LocalstackModels.LambdaSpec spec =
        new LocalstackModels.LambdaSpec(
            "fn",
            "python3.12",
            "handler.main",
            ".",
            List.of(new LocalstackModels.EnvPair("KEY", "VALUE")));
    ObjectNode config = new LocalstackDependencyBuilder("ls").withLambda(spec).build().forFfi();
    ObjectNode lambda = (ObjectNode) config.path("lambdas").get(0);
    assertEquals("fn", lambda.path("name").asText());
    assertEquals("python3.12", lambda.path("runtime").asText());
    assertEquals("handler.main", lambda.path("handler").asText());
    assertEquals(
        java.nio.file.Path.of(".").toAbsolutePath().normalize().toString(),
        lambda.path("source_dir").asText());
    ArrayNode env = (ArrayNode) lambda.path("environment");
    assertEquals("KEY", env.get(0).get(0).asText());
    assertEquals("VALUE", env.get(0).get(1).asText());
  }

  @Test
  void withLambda_defaultEnvironmentOverload_serializesEmptyEnvironment() {
    LocalstackModels.LambdaSpec spec =
        new LocalstackModels.LambdaSpec("fn", "python3.12", "handler.main", ".");
    ObjectNode config = new LocalstackDependencyBuilder("ls").withLambda(spec).build().forFfi();
    ArrayNode env = (ArrayNode) config.path("lambdas").get(0).path("environment");
    assertTrue(env.isEmpty());
  }

  @Test
  void withEventBus_appendsEventBusSpec() {
    ObjectNode config = new LocalstackDependencyBuilder("ls").withEventBus("orders").build().forFfi();
    assertEquals("orders", config.path("event_buses").get(0).path("name").asText());
  }

  @Test
  void withEventRule_sqsTarget_serializesSqsQueueTargetKind() {
    LocalstackModels.EventRuleSpec spec =
        new LocalstackModels.EventRuleSpec(
            "rule",
            "{\"source\":[\"custom\"]}",
            List.of(
                new LocalstackModels.EventRuleTarget(
                    "t1", new LocalstackModels.SqsQueueTarget("my-queue"))));
    ObjectNode config = new LocalstackDependencyBuilder("ls").withEventRule(spec).build().forFfi();
    ObjectNode rule = (ObjectNode) config.path("event_rules").get(0);
    assertEquals("rule", rule.path("name").asText());
    assertFalse(rule.has("event_bus"));
    ObjectNode target = (ObjectNode) rule.path("targets").get(0);
    assertEquals("t1", target.path("target_id").asText());
    assertEquals("sqs_queue", target.path("kind").asText());
    assertEquals("my-queue", target.path("queue_name").asText());
  }

  @Test
  void withEventRule_lambdaTargetWithEventBus_serializesLambdaTargetKind() {
    LocalstackModels.EventRuleSpec spec =
        new LocalstackModels.EventRuleSpec(
            "rule",
            "{\"source\":[\"custom\"]}",
            List.of(
                new LocalstackModels.EventRuleTarget(
                    "t1", new LocalstackModels.LambdaTarget("my-fn"))),
            "orders");
    ObjectNode config = new LocalstackDependencyBuilder("ls").withEventRule(spec).build().forFfi();
    ObjectNode rule = (ObjectNode) config.path("event_rules").get(0);
    assertEquals("orders", rule.path("event_bus").asText());
    ObjectNode target = (ObjectNode) rule.path("targets").get(0);
    assertEquals("lambda", target.path("kind").asText());
    assertEquals("my-fn", target.path("function_name").asText());
  }

  @Test
  void addChildDependency_nonEmptyChildren_serializesChildrenArray() {
    ObjectNode config =
        new LocalstackDependencyBuilder("ls")
            .addChildDependency(new StubDependency("child-a"))
            .build()
            .forFfi();
    ArrayNode children = (ArrayNode) config.path("children");
    assertEquals(1, children.size());
    assertEquals("child-a", children.get(0).path("identifier").asText());
  }

  @Test
  void identifier_returnsConfiguredIdentifier() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertTrue(dep.identifier().startsWith("arena-localstack-ls-"));
  }

  @Test
  void port_noExplicitPort_returnsInternalPortDefault() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(LocalstackModels.LOCALSTACK_INTERNAL_DOCKER_PORT, dep.port());
  }

  @Test
  void port_explicitPort_returnsConfiguredPort() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4599).build();
    assertEquals(4599, dep.port());
  }

  @Test
  void endpointUrl_defaultHost_usesLocalhost() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4566).build();
    assertEquals("http://localhost:4566", dep.endpointUrl());
  }

  @Test
  void endpointUrl_explicitHost_usesGivenHost() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4566).build();
    assertEquals("http://ls-host:4566", dep.endpointUrl("ls-host"));
  }

  @Test
  void internalEndpointUrl_noContainerName_usesIdentifierAsHost() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "http://" + dep.identifier() + ":" + LocalstackModels.LOCALSTACK_INTERNAL_DOCKER_PORT,
        dep.internalEndpointUrl());
  }

  @Test
  void internalEndpointUrl_explicitContainerName_usesContainerNameAsHost() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "http://ls-box:" + LocalstackModels.LOCALSTACK_INTERNAL_DOCKER_PORT,
        dep.internalEndpointUrl("ls-box"));
  }

  @Test
  void queueUrl_defaultHostAndAccount_buildsLocalstackStyleUrl() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4566).build();
    assertEquals(
        "http://localhost:4566/000000000000/my-queue", dep.queueUrl("my-queue"));
  }

  @Test
  void queueUrl_explicitHost_usesGivenHost() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4566).build();
    assertEquals(
        "http://ls-host:4566/000000000000/my-queue", dep.queueUrl("my-queue", "ls-host"));
  }

  @Test
  void queueUrl_explicitHostAndAccount_usesBothOverrides() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").withPort(4566).build();
    assertEquals(
        "http://ls-host:4566/111111111111/my-queue",
        dep.queueUrl("my-queue", "ls-host", "111111111111"));
  }

  @Test
  void queueArn_defaultRegionAndAccount_buildsArn() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:sqs:us-east-1:000000000000:my-queue", dep.queueArn("my-queue"));
  }

  @Test
  void queueArn_explicitRegion_usesGivenRegion() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:sqs:eu-west-1:000000000000:my-queue", dep.queueArn("my-queue", "eu-west-1"));
  }

  @Test
  void queueArn_explicitRegionAndAccount_usesBothOverrides() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:sqs:eu-west-1:222222222222:my-queue",
        dep.queueArn("my-queue", "eu-west-1", "222222222222"));
  }

  @Test
  void lambdaArn_defaultRegionAndAccount_buildsArn() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:lambda:us-east-1:000000000000:function:my-fn", dep.lambdaArn("my-fn"));
  }

  @Test
  void lambdaArn_explicitRegion_usesGivenRegion() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:lambda:eu-west-1:000000000000:function:my-fn",
        dep.lambdaArn("my-fn", "eu-west-1"));
  }

  @Test
  void lambdaArn_explicitRegionAndAccount_usesBothOverrides() {
    LocalstackDependency dep = new LocalstackDependencyBuilder("ls").build();
    assertEquals(
        "arn:aws:lambda:eu-west-1:333333333333:function:my-fn",
        dep.lambdaArn("my-fn", "eu-west-1", "333333333333"));
  }
}
